### ISSUE TITLE
Name of columns in subqueries with unions

### ISSUE THREAD

**Issue by theosotr** [sqlalchemy/sqlalchemy#5399]
**Describe your question**

I have the following query:

```python
q1 = session.query(Model.col1, Model.col2).select_from(Model1)
q2 = session.query(Model2.col3,Model.col4).select_from(Model2)
subq1 = q1.union(q2).subquery()
subq1.c.col1 # Works fine
```
but when I have this query
```python
q3 = q1.union(q2)
subq2 = q3.union(session.query(Model2.col3,Model.col4).select_from(Model2)).subquery()
subq2.c.col1
```
I get  `*** AttributeError: col1`.

Notably, when I debug the program I get the following
```bash
(Pdb) subq2.c._all_columns
[<sqlalchemy.sql.elements.ColumnClause at 0x7f8baf5401d0; %(140237919345632 anon)s_col1>, <sqlalchemy.sql.elements.ColumnClause at 0x7f8baf5402e8; %(140237919345632 anon)s_col2>]
```

The subquery does not contain the attributes as I expected. Apparently I am missing something here, and I don't get a subquery with the expected attributes, that is the name of the column in subquery is `%(140237919345632 anon)s_col1>` and not `s_col1`)

**Have a nice day!**

**Comment by zzzeek** [comment]
so you are in the darkest area of the ORM's API and you're in fact in a place that is part of why I'm having select() take over for Query in 1.4 / 2.0.    the way that Query.union() creates a subquery (before you actualy called subquery(), it did) as well as that it names columns with tablenames is not helpful to anyone when trying to construct objects of this nature.    1.4 also includes refinements to how these labels are made.

I think your best bet for building these queries is to just use select() directly, and use the union() standalone to gather them up.  select() even in 1.3 supports ORM attributes and things like that.     the only thing 1.3's select() won't get you is the query.join() method, you would have to do stmt.select_from(orm_join(Entity1, Entity2)) for that.

**Comment by theosotr** [comment]
The strange thing here was that I was building a subquery to which I applied union() two times, i.e.,

```python
tmp = q1.union(q2)
subq = tmp.union(q3).subquery()
```
if it is the other way round, I was correctly getting the attributes of q3 in subquery `subq`
```python
tmp = q1.union(q2)
subq = q3.union(tmp).subquery()
```

So from understanding, there are refinements in 1.4 / 2.0 to support this kind of queries, right?
Could you please provide a concrete example demonstrating how to build such a query in 1.3?

**Comment by zzzeek** [comment]
yes can you show me the models and exactly the SQL you are looking for?   is this a straight union of three select() statements or are you trying to have nesting?

**Comment by zzzeek** [comment]
e.g. what are the models here, there seem to be three:

q1 = session.query(Model.col1, Model.col2).select_from(Model1)
q2 = session.query(Model2.col3,Model.col4).select_from(Model2)

do you really mean to have "Model2.col3" and "Model.col4" in the second query like that?  same with selecitng from "Model.x / Model.y" but then select_from(Model1).

**Comment by theosotr** [comment]
I have three arbitrary queries, namely, q1, q2, q3. Consider the following simplified example

```python
q1 = session.query(Simplex.id.label('foo')).select_from(Simplex)
q2 = session.query(Simplex.id.label('foo')).select_from(Simplex)
q3 = session.query(Simplex.id.label('foo')).select_from(Simplex)

tmp = q1.union(q2)
sub = tmp.union(q3).subquery()

ret = session.query(func.sum(sub.c.foo)).select_from(sub).firs
print(ret)
```
but I am getting

```bash
Traceback (most recent call lst):
  File "/dir/.env/lib/python3.6/site-packages/SQLAlchemy-1.4.0b1.dev0-py3.6-linux-x86_64.egg/sqlalchemy/sql/base.py", line 975, in __getattr__
    return self._index[key]
KeyError: 'foo'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "driver_sqlite.py", line 36, in <module>
    ret = session.query(func.sum(sub.c.id)).select_from(sub)
  File "/dir/.env/lib/python3.6/site-packages/SQLAlchemy-1.4.0b1.dev0-py3.6-linux-x86_64.egg/sqlalchemy/sql/base.py", line 977, in __getattr__
    util.raise_(AttributeError(key), replace_context=err)
  File "/dir/.env/lib/python3.6/site-packages/SQLAlchemy-1.4.0b1.dev0-py3.6-linux-x86_64.egg/sqlalchemy/util/compat.py", line 169, in raise_
    raise exception
AttributeError: foo
```

I want to construct the SQL query below

```sql
SELECT sum(anon_1.anon_2_anon_3_foo) AS sum_1
FROM (
  SELECT anon_2.anon_3_foo AS anon_2_anon_3_foo
  FROM (
    SELECT anon_3.foo AS anon_3_foo
    FROM (
      SELECT simplex.id AS foo
      FROM simplex
      UNION SELECT simplex.id AS foo
      FROM simplex
     ) AS anon_3
     UNION SELECT simplex.id AS foo
     FROM simplex
   ) AS anon_2
) AS anon_1
```

**Comment by zzzeek** [comment]
so what happens here is that Query really wants to create these <tablename>_<colname> labels and you need to make sure it doesn't do that.  especially because <tablename> here is an anonymous alias.  when we take "tmp" and print it, it is already doing the aliasing that will mess up the labels:

```
(Pdb) print(tmp)
SELECT anon_1.foo AS anon_1_foo 
FROM (SELECT a.id AS foo 
FROM a UNION SELECT a.id AS foo 
FROM a) AS anon_1

```

that "anon_1_foo" label is going to get put into the next query and that's where it all goes wrong.

so you want to break out of Query-ness like this:

```
sub = union(tmp.statement, q3).alias() 
```

then the re-labeling does not happen:


```
(Pdb) print(sub)
SELECT anon_1.foo 
FROM (SELECT a.id AS foo 
FROM a UNION SELECT a.id AS foo 
FROM a) AS anon_1 UNION SELECT a.id AS foo 
FROM a

```

then you can continue with that to get a statement that's equivalent to what you are doing:
```

q = session.query(func.sum(sub.c.foo)).select_from(sub)
print(q)
```

```
SELECT sum(anon_1.foo) AS sum_1 
FROM (SELECT anon_2.foo AS foo 
FROM (SELECT a.id AS foo 
FROM a UNION SELECT a.id AS foo 
FROM a) AS anon_2 UNION SELECT a.id AS foo 
FROM a) AS anon_1

```


if you just go with Core to avoid all the weird labeling, this is what SQLAlchemy 2.0 will mostly look like too
```

q1 = select([Simplex.id.label('foo')]).select_from(Simplex)
q2 = select([Simplex.id.label('foo')]).select_from(Simplex)
q3 = select([Simplex.id.label('foo')]).select_from(Simplex)


tmp = union(q1, q2)

sub = union(tmp, q3).alias()

q = session.query(func.sum(sub.c.foo)).select_from(sub)
print(q)
```

**Comment by theosotr** [comment]
I see. Great; I really appreciate your help.

**Comment by zzzeek** [comment]
ok great.   query.union() works kind of wrong, basically.  this is a big part of why I'm moving towards select() being the one way to do everything.  the tablename_columnname labeling thing is not actually needed in 1.4 either as result set mechanics work positionally now.

**Comment by zzzeek** [comment]
Indeed based on this [tweet](https://twitter.com/CoolSWEng/status/1352655487624486912) and then observing issues raised by @theosotr  I would assume these issues have all come from the Cynthia project.   So to that extent we can say that the poor behavior here is a bug that's fixed by #5159 which allows the select() construct to be used, which has a better API for things like nested unions.
